### PR TITLE
Fix memory leaks

### PR DESCRIPTION
**Pull request by burelc-amzn**
There's a few different leaks being fixed by this PR, check the individual commits for the details

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h [hunk 1]
```diff
@@ -206,25 +216,48 @@ namespace AZ
...
+                // BehaviorValueParameter does not destroy instances of types set
+                // with StoreInTempData(). It does not take ownership of that
+                // instance, because it is designed to reference existing values on
+                // the stack. For a parameter's default value, however, that
+                // instance must be owned by something, so that ownership is the
+                // responsibility of the BehaviorDefaultValue type. If the type of
+                // the default value has a non-trivial destructor, this class will
+                // invoke that destructor when it itself is destroyed.
+                AZStd::destroy_at(reinterpret_cast<AZStd::decay_t<Value>*>(m_value.m_value));
+            }
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:255]: This also check `m_value.m_value` is not nullptr before invoking the destructor on it.
Someone could create a default value parameter with a value that is nullptr.
Such as `behaviorContext->MakeDefaultValue(static_cast<int*>(nullptr))`

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h [hunk 2]
```diff
@@ -206,25 +216,48 @@ namespace AZ
...
+    {
+    public:
         /**
         * Create a default value for a specific method parameter. The Default values is stored by value
         * in a temp storage, so currently there is limit the \ref BehaviorValueParameter temp storage, we
         * can easily change that if it became a problem.
         */
-        template<typename Value>
-        BehaviorDefaultValue(Value&& value)
+        BehaviorDefaultValueTyped(Value&& value)
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:237]: This is an rvalue only constructor.
It is a forwarding constructor as the `Value` type is not a template type in regard to this function

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

File: Code/Framework/AzCore/AzCore/Serialization/ObjectStream.cpp [hunk 3]
```diff
@@ -1115,6 +1115,10 @@ namespace AZ
                     if (GenericClassInfo* genericClassInfo = sc.FindGenericClassInfo(cd->m_typeId))
                     {
                         element.m_id = genericClassInfo->GetSpecializedTypeId();
+                        if (genericClassInfo->GetGenericTypeId() == GetAssetClassId())
+                        {
+                            element.m_id = GetAssetClassId();
+                        }
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Serialization/ObjectStream.cpp:1121]: This code shouldn't swap out the `AZ::Data::Asset<AssetDataDerivedClass>` aggregate Id for the generic AssetClassId, instead it should update the check in [ObjectStream::LoadClass](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Serialization/ObjectStream.cpp#L769) to query the GenericClassInfo from the SerializeContext there and determine if the the aggregate id maps to the genericAssetClassId

File: Code/Framework/AzCore/AzCore/Serialization/ObjectStream.cpp [hunk 4]
```diff
@@ -1242,6 +1246,10 @@ namespace AZ
...
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Serialization/ObjectStream.cpp:1252]: Repeat: The `element.m_id` should remain set to the specialized type id and the LoadClass check for `if (element.m_id == GetAssetClassId())` should be updated to check the genericClassInfo.

```
if (GenericClassInfo* genericClassInfo = m_sc->FindGenericClassInfo(element.m_id);
    genericClassInfo && genericClassInfo->GetGenericTypeId() == GetAssetClassId())
{
   // ....
```

File: Code/Framework/AzCore/AzCore/Serialization/ObjectStream.cpp [hunk 5]
```diff
@@ -1350,6 +1358,10 @@ namespace AZ
...
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Serialization/ObjectStream.cpp:1364]: Repeat: The DataElement type id should be set to the specialized type id and the check in LoadClass for the Generic Asset Class Id should be updated to see if the specialized type has generic class info with a generic type id that matches the Generic Asset Class Id

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyEditorAPI_Internals.h [hunk 6]
```diff
@@ -232,7 +234,19 @@ namespace AzToolsFramework
...
+                {
+                    // Look up the reflected SerializeContext class data in order to find the RttiHelper
+                    const AZ::SerializeContext::ClassData* classData = serializeContext->FindClassData(id);
+                    const AZ::IRttiHelper* rttiHelper = classData != nullptr ? classData->m_azRtti : nullptr;
+                    return id == AZ::GetAssetClassId() ||
+                        (rttiHelper != nullptr && rttiHelper->GetGenericTypeId() == AZ::GetAssetClassId());
+                }
+            }
+
+            return false;
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyEditorAPI_Internals.h:249]: This still should check that the `AZ::SerializeTypeInfo<PropertyType>::GetUuid() == id`

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

File: Gems/Blast/Code/Source/Components/BlastFamilyComponent.cpp [hunk 7]
```diff
@@ -71,33 +71,33 @@ namespace Blast
                     "Radial Damage", &BlastFamilyDamageRequests::RadialDamage,
                     {{{"position", "The global position of the damage's hit."},
                       {"minRadius", "Damages all chunks/bonds that are in the range [0, minRadius] with full damage.",
-                       AZ::BehaviorDefaultValuePtr(aznew AZ::BehaviorDefaultValue(static_cast<float>(0.0)))},
+                       AZ::BehaviorDefaultValuePtr(aznew AZ::BehaviorDefaultValueTyped(static_cast<float>(0.0)))},
```
Comment (lumberyard-employee-dm) [inline Gems/Blast/Code/Source/Components/BlastFamilyComponent.cpp:74]: These functions can be switched to just use the `behaviorContext->MakeDefaultValue` function.
That way the `AZ::BehaviorDefaultTyped` class never needs to be interacted with directly.

File: cmake/Platform/Common/Clang/Configurations_clang.cmake [hunk 8]
```diff
@@ -57,5 +57,15 @@ ly_append_configurations_options(
...
+if(LY_BUILD_WITH_ADDRESS_SANITIZER)
+    ly_append_configurations_options(
+        COMPILATION_DEBUG
+            -fsanitize=address
+            -fno-omit-frame-pointer
+        LINK_NON_STATIC_DEBUG
+            -shared-libsan
+            -fsanitize=address
+    )
+endif()
```
Comment (lumberyard-employee-dm) [inline cmake/Platform/Common/Clang/Configurations_clang.cmake:69]: Is the optimization in profile configuration not worth supporting the address sanitizer?

File: Code/Framework/AzCore/AzCore/Asset/AssetSerializer.h [hunk 9]
```diff
@@ -120,7 +120,7 @@ namespace AZ {
 
             const Uuid& GetSpecializedTypeId() const override
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/Asset/AssetSerializer.h:121]: question: Should these changes be in a separate PR? Were they somehow contributing to a leak? This is something @AMZN-stankowi should likely take a look at

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h [hunk 10]
```diff
@@ -206,25 +216,53 @@ namespace AZ
...
+                    m_value.m_value = nullptr;
+                }
+            }
         }
-
-        BehaviorValueParameter m_value;
     };
 
+    template<typename Value>
+    BehaviorDefaultValueTyped(Value) -> BehaviorDefaultValueTyped<AZStd::decay_t<Value>>;
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:265]: suggestion: Might be worth adding a comment this is a deduction guide

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h [hunk 11]
```diff
@@ -2262,8 +2300,14 @@ namespace AZ
     }
 
     //////////////////////////////////////////////////////////////////////////
-    template<typename T>
+    template<typename T, typename>
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:2303]: question: Sorry if this is a stupid question but why is this unused `typename` needed? Might be worth a comment as it is definitely a bit odd to see if one isn't familiar with what it's required for

File: Gems/PhysX/Code/Source/System/PhysXSystem.cpp [hunk 12]
```diff
@@ -64,8 +64,8 @@ namespace PhysX
         m_onMaterialLibraryReloadedCallback(asset);
     }
 
-    PhysXSystem::PhysXSystem(PhysXSettingsRegistryManager* registryManager, const physx::PxCookingParams& cookingParams)
-        : m_registryManager(*registryManager)
+    PhysXSystem::PhysXSystem(AZStd::unique_ptr<PhysXSettingsRegistryManager> registryManager, const physx::PxCookingParams& cookingParams)
+        : m_registryManager(AZStd::move(registryManager))
```
Comment (hultonha) [inline Gems/PhysX/Code/Source/System/PhysXSystem.cpp:68]: @SergeyAMZN @greerdv @moraaar - worth sanity checking this, looks good to me but would be good to get confirmation

**Review by hultonha — COMMENTED** [review]
Overall looks good to me, I have a couple of questions and looped in a few other reviewers to take a look, once they have signed off I'd be happy to approve 👍 

Really awesome to have added this! 🙂 Great work!

File: Gems/PhysX/Code/Source/System/PhysXSystem.cpp (hunk 12 above)
Comment (SergeyAMZN) [reply Gems/PhysX/Code/Source/System/PhysXSystem.cpp:68]: Looks good to me. 
Interestingly I do remember seeing this memory leak before! 
Either another PR that never made development or just a leaks report somewhere..

**Review by SergeyAMZN — APPROVED** [review]
Gems/PhysX changes look good

File: Code/Framework/AzCore/AzCore/Asset/AssetSerializer.h (hunk 9 above)
Comment (burelc-amzn) [reply Code/Framework/AzCore/AzCore/Asset/AssetSerializer.h:121]: Yeah, you can take a look at the commit message of d10096442ba68bcdecc0fc68ab83954b9303fb0b for the full details. The TL;DR; is the old code stored a map where the keys are typeids. It would use one key to check if the map contained a typeid, and if it didn't, it used a different typeid to store an object in the map. The key used to store the item was the same key for many different lookups, so the value stored at that key would get overwritten. Because the value type is a raw pointer, the overwritten pointer would cause a leak.

This fixes the leak by making the key used for lookup and the key used for insertion the same.

As for whether this should be in a separate PR, I don't think it matters. This change was done in its own commit, and I plan to merge it with a merge commit (not squash), so if this one commit causes issues down the line, the commit can be reverted independently of the rest of the changes from this PR.

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h (hunk 11 above)
Comment (burelc-amzn) [reply Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:2303]: We're already in kind of an unusual situation here, since this is an out-of-line definition of a templated member function. If I had been writing this from the get-go, I would've left the definition inline. Once you jump to the [declaration of the function](https://github.com/o3de/o3de/blob/1d0cbe723d84c17eabbbefde6ba5d32471c120a9/Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h#L193-L194), then the meaning of that second `typename` argument becomes clear:

```cpp
/// Used internally to store values in the temp data
template<typename T, typename = AZStd::enable_if_t<AZStd::is_trivially_destructible_v<AZStd::decay_t<T>>>>
void StoreInTempData(T&& value);
```

That second template parameter has a default value, using `enable_if` to prevent it from being called with types that are not trivially destructible.

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h [hunk 13]
```diff
@@ -206,25 +216,53 @@ namespace AZ
...
+    public:
         /**
         * Create a default value for a specific method parameter. The Default values is stored by value
         * in a temp storage, so currently there is limit the \ref BehaviorValueParameter temp storage, we
         * can easily change that if it became a problem.
         */
-        template<typename Value>
-        BehaviorDefaultValue(Value&& value)
+        template<typename U, typename = AZStd::enable_if_t<AZStd::is_same_v<Value, AZStd::decay_t<U>>>>
+        BehaviorDefaultValueTyped(U&& value)
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:238]: Why have an enable_if_t check that is restrictive?
The only check that should be here is to make sure that U isn't derived from type `BehaviorDefaultValue`.

Besides that any U should be supported as long as it is convertible to `Value`
```suggestion
        template<typename U, typename = AZStd::enable_if_t<AZStd::conjunction_v<
        !AZStd::derived_from<AZStd::decay_t<U>, BehaviorDefaultValue>,
        AZStd::convertible_to<U, Value>>>
        BehaviorDefaultValueTyped(U&& value)
```

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h (hunk 11 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:2303]: Well this is the definition to of a template function with a default typename value used for SFINAE [above](https://github.com/o3de/o3de/pull/9279/files/1d0cbe723d84c17eabbbefde6ba5d32471c120a9#diff-a8f55fdb89c13916025263fc92d7e9aee05825950673c641c3dddf183f50f347R193)

Comment (burelc-amzn) [reply Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:2303]: And just for further clarification, calling `StoreInTempData` with a type that has a user-defined destructor risks leaking memory. `StoreInTempData` creates a copy of the `value` parameter, takes ownership of the copy, but will not invoke that copy's destructor. This means that [a type like `AZ::SettingsRegistryScriptUtils::Internal::SpecializationsProxy`](https://github.com/o3de/o3de/blob/6264fa97c9f55fe00d76a09e4ed09814be020a98/Code/Framework/AzCore/AzCore/Settings/SettingsRegistryScriptUtils.cpp#L57-L81), which is copyable but stores a `unique_ptr`, would not have its destructor run to delete the value that its `unique_ptr` holds.

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h [hunk 14]
```diff
@@ -2898,7 +2942,7 @@ namespace AZ
     template<class Value>
     BehaviorDefaultValuePtr BehaviorContext::MakeDefaultValue(Value&& defaultValue)
     {
-        return !IsRemovingReflection() ? aznew BehaviorDefaultValue(AZStd::forward<Value>(defaultValue)) : nullptr;
+        return !IsRemovingReflection() ? aznew BehaviorDefaultValueTyped(AZStd::forward<Value&&>(defaultValue)) : nullptr;
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:2945]: This is incorrect, forward should pass `Value`, not `Value&&`.
This has a slight change when supplying an rvalue reference to a function.
By default a temporary or `std::move` value deduces Value as `T` not `T&&`.
std::forward expects to be supplied a `T` or a const qualified `T&`, where it then adds an rvalue reference to the type

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h (hunk 13 above)
Comment (burelc-amzn) [reply Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:238]: I tried that, but changing `BehaviorValueParameter` to correctly destroy the type it stores depending on if `StoreInTempData()` was used proved to be really tricky. Given the current use of this type, `BehaviorValueParameter` needs to be copyable and movable, which is not possible because it uses a stack-based allocator to allocate space to store the instance.

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:238]: This is not changing the BehaviorValueParameter, this is changing the BehaviorDefaultValue class which encapsulates a `BehaviorValueParameter`
Having a derived `BehaviorDefaultValueTyped` doesn't change stack based allocator issue of the BehaviorValueParameter

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h (hunk 14 above)
Comment (burelc-amzn) [reply Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:2945]: Curse you C++ and your overly complex reference collapsing and type deduction rules.

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h (hunk 10 above)
Comment (burelc-amzn) [reply Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:265]: With @lumberyard-employee-dm 's suggestion below, I'm removing this class anyway, so the deduction guide is gone too.

File: Code/Framework/AzCore/AzCore/Asset/AssetSerializer.h (hunk 9 above)
Comment (hultonha) [reply Code/Framework/AzCore/AzCore/Asset/AssetSerializer.h:121]: Aaah I see interesting! Thank you for the explanation, much appreciated. Sure thing makes sense :+1: I just wasn't sure on the surface if it was related, but it clearly is! 😅 Thanks!

File: Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h (hunk 11 above)
Comment (hultonha) [reply Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h:2303]: I see thanks for the explanation, I am a bit rusty on the old template meta-programming these 🙈 need to brush up to remember these parts, thank you for indulging me 🙂

**Review by lawsonamzn — APPROVED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]

**Review by lawsonamzn — APPROVED** [review]

**Review by lawsonamzn — APPROVED** [review]
Nice.  Const Shared_ptr<>& is always a bit of a code smell, I think.

**Review by lawsonamzn — APPROVED** [review]

File: Code/Tools/AssetProcessor/AssetBuilderSDK/AssetBuilderSDK/SerializationDependencies.cpp [hunk 15]
```diff
@@ -24,16 +24,16 @@ namespace AssetBuilderSDK
         {
             return false;
         }
-        if (classData->m_typeId == AZ::GetAssetClassId())
+        const auto isAssetClass = [&serializeContext, classData]
         {
-            auto* asset = reinterpret_cast<AZ::Data::Asset<AZ::Data::AssetData>*>(instancePointer);
-
-            if (asset->GetId().IsValid())
+            if (classData->m_typeId == AZ::GetAssetClassId() && classData->m_version == 1)
```
Comment (lumberyard-employee-dm) [inline Code/Tools/AssetProcessor/AssetBuilderSDK/AssetBuilderSDK/SerializationDependencies.cpp:29]: This needs to be `classData->m_version <=2` as version 2 `AZ::Data::Asset` class serialization writes out the AssetClassId

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]

**Review by hultonha — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyManagerComponent.cpp [hunk 16]
```diff
@@ -263,19 +263,19 @@ namespace AzToolsFramework
...
                 {
-                    return pHandlerFound;
+                    pHandlerFound = ResolvePropertyHandler(handlerName, (*cls)->m_typeId);
+                    if (pHandlerFound)
+                    {
+                        return pHandlerFound;
+                    }
                 }
-                else
+                if (const auto* genericInfo = sc->FindGenericClassInfo(handlerType))
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyManagerComponent.cpp:274]: This needs to be specialized for `GetAssetClassId`

Comment (burelc-amzn) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyManagerComponent.cpp:274]: That would only happen if someone registered a handler with the id equal to whatever `GetGenericClassInfoVectorTypeId()` returns. That actually seems like it could be useful in other contexts. Or should we just keep on special casing `GetAssetClassId()`?

Comment (lumberyard-employee-dm) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyManagerComponent.cpp:274]: `FindGenericClassInfo` accepts the template specialization ID, not the generic ID. Passing in the GenericClassInfo TypeId wouldn't ever locate the genericInfo.

The if block currently only works for the `GetAssetClassId()` scenario, because of the special case code in AssetSerializer.h to register [GenericClassInfo](https://github.com/o3de/o3de/pull/9279/files#diff-f04464db71c462bd0ce78e7af436e1c08c7170e88d749e736eee4e038f7bb978R177) using the `GetAssetClassId()`.

This code should be fine.

**Review by moraaar — APPROVED** [review]
Changes to PhysX an Blast gems look good. Thanks!

**Comment by burelc-amzn** [comment]
The commit to fix the memory leak for the Asset handlers isn't ready yet (tests are failing), so I've removed it from this PR. The rest are unchanged.

**Review by lawsonamzn — APPROVED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]
